File: analysis/ghidra/DumpProtocol.java
// Ghidra headless: locate functions referencing protocol strings, decompile them.
// @category Custom

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Data;
import ghidra.program.model.listing.DataIterator;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.symbol.Reference;
import ghidra.program.model.symbol.ReferenceIterator;
import ghidra.program.model.symbol.ReferenceManager;
import ghidra.util.task.ConsoleTaskMonitor;

import java.io.PrintWriter;
import java.io.FileWriter;
import java.util.*;

public class DumpProtocol extends GhidraScript {

    static final String OUT = "/Users/alexscott/Projects/LEDctrl/analysis/ghidra/decomp.txt";
    static final String[] NEEDLES = {
        "SEQUENT.SYS", "temp.Nmg", "uJetFileICMD.pas", "uJetFileIICMD.pas",
        "RUNTIME.SYS", "CONFIG.SYS", "JetFileII", "SequentList.tmps",
        "MakeNmg", "Nmg", "TJetFIICMD", "TJetFICMD", "TBaseCMD"
    };

    @Override
    protected void run() throws Exception {
        Listing listing = currentProgram.getListing();
        ReferenceManager rm = currentProgram.getReferenceManager();

        Map<String, List<Address>> hits = new LinkedHashMap<>();
        for (String n : NEEDLES) hits.put(n, new ArrayList<>());

        DataIterator it = listing.getDefinedData(true);
        while (it.hasNext()) {
            Data d = it.next();
            try {
                Object v = d.getValue();
                if (v == null) continue;
                String s = v.toString();
                for (String n : NEEDLES) {
                    if (s.contains(n)) hits.get(n).add(d.getAddress());
                }
            } catch (Exception e) {}
        }

        Map<String, Function> fcns = new TreeMap<>();
        Map<String, Set<String>> fcnNeedles = new TreeMap<>();

        for (Map.Entry<String, List<Address>> e : hits.entrySet()) {
            for (Address a : e.getValue()) {
                ReferenceIterator refs = rm.getReferencesTo(a);
                while (refs.hasNext()) {
                    Reference r = refs.next();
                    Function f = currentProgram.getFunctionManager()
                        .getFunctionContaining(r.getFromAddress());
                    if (f == null) continue;
                    String ep = f.getEntryPoint().toString();
                    fcns.put(ep, f);
                    fcnNeedles.computeIfAbsent(ep, k -> new TreeSet<>()).add(e.getKey());
                }
            }
        }

        DecompInterface di = new DecompInterface();
        di.openProgram(currentProgram);
        ConsoleTaskMonitor mon = new ConsoleTaskMonitor();

        try (PrintWriter fh = new PrintWriter(new FileWriter(OUT))) {
            fh.println("=== Strings located ===");
            for (Map.Entry<String, List<Address>> e : hits.entrySet()) {
                if (e.getValue().isEmpty()) continue;
                fh.println(e.getKey() + ": " + e.getValue());
            }
            fh.println();
            fh.println("=== Functions referencing them ===");
            for (Map.Entry<String, Function> e : fcns.entrySet()) {
                fh.println(e.getKey() + "  " + e.getValue().getName()
                    + "  refs=" + fcnNeedles.get(e.getKey()));
            }
            fh.println();
            fh.println("=== Decompiled ===");
            for (Map.Entry<String, Function> e : fcns.entrySet()) {
                Function f = e.getValue();
                fh.println();
                fh.println(";;; ============================================");
                fh.println(";;; " + e.getKey() + "  " + f.getName()
                    + "  refs=" + fcnNeedles.get(e.getKey()));
                fh.println(";;; ============================================");
                try {
                    DecompileResults r = di.decompileFunction(f, 60, mon);
                    if (r != null && r.getDecompiledFunction() != null) {
                        fh.print(r.getDecompiledFunction().getC());
                    } else {
                        fh.println("// decompile failed");
                    }
                } catch (Exception ex) {
                    fh.println("// exception: " + ex);
                }
            }
        }
        println("DONE -> " + OUT);
    }
}

File: analysis/ghidra/dump_protocol.py
# Ghidra headless script: find functions referencing SEQUENT.SYS / temp.Nmg / Pascal source paths
# and dump their decompilation to a file.
# @category Custom

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor
import os

OUT = "/Users/alexscott/Projects/LEDctrl/analysis/ghidra/decomp.txt"

def find_string_addrs(needles):
    out = {}
    listing = currentProgram.getListing()
    mem = currentProgram.getMemory()
    data = listing.getDefinedData(True)
    for d in data:
        try:
            v = d.getValue()
            if v is None:
                continue
            s = str(v)
            for n in needles:
                if n in s:
                    out.setdefault(n, []).append((d.getAddress(), s))
        except:
            pass
    return out

def get_xrefs(addr):
    refs = currentProgram.getReferenceManager().getReferencesTo(addr)
    return [r.getFromAddress() for r in refs]

def fcn_for(addr):
    return currentProgram.getFunctionManager().getFunctionContaining(addr)

di = DecompInterface()
di.openProgram(currentProgram)
mon = ConsoleTaskMonitor()

needles = ["SEQUENT.SYS", "temp.Nmg", "uJetFileICMD.pas", "uJetFileIICMD.pas",
           "RUNTIME.SYS", "CONFIG.SYS", "JetFileII", "SequentList.tmps"]

found = find_string_addrs(needles)

fcns = {}
for n, hits in found.items():
    for addr, s in hits:
        for caller in get_xrefs(addr):
            f = fcn_for(caller)
            if f:
                fcns.setdefault(f.getEntryPoint().toString(), (f, set())) [1].add(n)

with open(OUT, "w") as fh:
    fh.write("=== Strings located ===\n")
    for n, hits in sorted(found.items()):
        fh.write(f"{n}: {[a.toString() for a,_ in hits]}\n")
    fh.write("\n=== Functions referencing them ===\n")
    for ep, (f, ns) in sorted(fcns.items()):
        fh.write(f"{ep}  {f.getName()}  refs={sorted(ns)}\n")
    fh.write("\n=== Decompiled ===\n")
    for ep, (f, ns) in sorted(fcns.items()):
        fh.write(f"\n\n;;; ============================================\n")
        fh.write(f";;; {ep}  {f.getName()}  refs={sorted(ns)}\n")
        fh.write(f";;; ============================================\n")
        try:
            r = di.decompileFunction(f, 60, mon)
            if r and r.getDecompiledFunction():
                fh.write(r.getDecompiledFunction().getC())
            else:
                fh.write("// decompile failed\n")
        except Exception as e:
            fh.write(f"// exception: {e}\n")

print("DONE -> %s" % OUT)
